repo/blob/gcs/gcs_pit.go

```diff
@@ -0,0 +1,174 @@
+	fmt.Printf("getMetadata del pit %s\n", b)
+	var vml []versionMetadata
+
+	if err := gcs.getBlobVersions(ctx, b, func(m versionMetadata) error {
+		fmt.Printf("  trovata versione %s del %s. deleted=%t\n", m.Version, m.Timestamp, m.IsDeleteMarker)
+		// only include versions older than s.pointInTime
+		if !m.Timestamp.After(gcs.pointInTime) {
+			fmt.Printf("      Skippo versione %s del %s. Troppo nuova.\n", m.Version, m.Timestamp)
+			vml = append(vml, m)
+		}
+
+		return nil
+	}); err != nil {
+		return versionMetadata{}, errors.Wrapf(err, "could not get version metadata for blob %s", b)
+	}
+
+	if v, found := newestAtUnlessDeleted(vml, gcs.pointInTime); found {
+		return v, nil
+	}
+
+	return versionMetadata{}, blob.ErrBlobNotFound
+}
+
+// newestAtUnlessDeleted returns the last version in the list older than the PIT.
+func newestAtUnlessDeleted(vs []versionMetadata, t time.Time) (v versionMetadata, found bool) {
+	fmt.Printf("newestAtUnlessDeleted %s\n", t)
+
+	for _, xxx := range vs {
+		fmt.Printf("newestAtUnlessDeleted: PRE SORT: found version %s del %s. deleted=%t\n", xxx.Version, xxx.Timestamp, xxx.IsDeleteMarker)
+	}
+
+	// Sort
+	sort.Slice(vs, func(i, j int) bool {
+		return vs[i].Timestamp.Before(vs[j].Timestamp)
+	})
+
+	for _, xxx := range vs {
+		fmt.Printf("newestAtUnlessDeleted POST SORT: found version %s del %s. deleted=%t\n", xxx.Version, xxx.Timestamp, xxx.IsDeleteMarker)
+	}
+
+	vs = getOlderThan(vs, t)
+
+	if len(vs) == 0 {
+		return versionMetadata{}, false
+	}
+
+	//v = vs[0] // versione s3
+	v = vs[len(vs)-1] // versione azure
+
+	fmt.Printf("newestAtUnlessDeleted del pit %s deleted=%t\n", v.Timestamp, v.IsDeleteMarker)
+
+	//return v, !v.IsDeleteMarker
+	return v, true
+}
+
+// Removes versions that are newer than t. The filtering is done in place
+// and uses the same slice storage as vs. Assumes entries in vs are in descending
+// timestamp order.
+func getOlderThan(vs []versionMetadata, t time.Time) []versionMetadata {
+	fmt.Printf("getOlderThan %s \n", t)
+
+	sort.Slice(vs, func(i, j int) bool { return vs[i].Timestamp.Before(vs[j].Timestamp) })
+
+	for i := range vs {
+		fmt.Printf("getOlderThan %s: vs[i].Timestamp: %s \n", t, vs[i].Timestamp)
+		if vs[i].Timestamp.After(t) {
+			fmt.Printf("  Skippo versione %s del %s. Troppo nuova.\n", vs[i].Version, vs[i].Timestamp)
+			return vs[i:]
+		}
+	}
+
+	return vs
+}
+
+// maybePointInTimeStore wraps s with a point-in-time store when s is versioned
+// and a point-in-time value is specified. Otherwise s is returned.
+func maybePointInTimeStore(ctx context.Context, gcs *gcsStorage, pointInTime *time.Time) (blob.Storage, error) {
+	if pit := gcs.Options.PointInTime; pit == nil || pit.IsZero() {
+		return gcs, nil
+	}
+
+	// Does the bucket supports versioning?
+	attrs, err := gcs.bucket.Attrs(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get determine if bucket '%s' supports versioning", gcs.BucketName)
+	}
+
+	if !attrs.VersioningEnabled {
```

Comment:
We also should error if ObjectRetentionMode != "Enabled"
Otherwise connecting/creating to a repo fails: `ERROR cannot initialize repository: unable to write blobcfg blob: PutBlob() failed for "kopia.blobcfg": unable to complete PutBlob(kopia.blobcfg) despite 10 retries: unexpected GCS error: googleapi: Error 400: Bucket does not have object retention enabled., invalid`